release-content/migration-guides/composable_specialization.md

```diff
@@ -0,0 +1,153 @@
+    fn from_world(&mut World) -> Self {
+        ...
+    }
+}
+
+// before
+impl SpecializedRenderPipeline for MySpecializer {
+    type Key = MyKey;
+
+    fn specialize(&self, key: Self::Key) -> RenderPipelineDescriptor {
+        RenderPipelineDescriptor {
+            label: Some("my_pipeline".into()),
+            layout: vec![
+                if key.msaa.samples() > 0 {
+                    self.layout_msaa.clone()
+                } else { 
+                    self.layout.clone() 
+                }
+            ],
+            push_constant_ranges: vec![],
+            vertex: VertexState {
+                shader: self.vertex.clone(),
+                shader_defs: vec![],
+                entry_point: "vertex".into(),
+                buffers: vec![],
+            },
+            primitive: Default::default(),
+            depth_stencil: None,
+            multisample: MultisampleState {
+                count: key.msaa.samples(),
+                ..Default::default()
+            },
+            fragment: Some(FragmentState {
+                shader: self.fragment.clone(),
+                shader_defs: vec![],
+                entry_point: "fragment".into(),
+                targets: vec![Some(ColorTargetState {
+                    format: TextureFormat::Rgba8Unorm,
+                    blend: Some(key.blend_state),
+                    write_mask: ColorWrites::all(),
+                })],
+            }),
+            zero_initialize_workgroup_memory: false,
+        },
+    }
+}
+
+app.init_resource::<SpecializedRenderPipelines<MySpecializer>>();
+
+// after
+impl Specialize<RenderPipeline> for MySpecializer {
+    type Key = MyKey;
+
+    fn specialize(
+        &self,
+        key: Self::Key,
+        descriptor: &mut RenderPipeline,
+    ) -> Result<Canonical<Self::Key>, BevyError> {
+        descriptor.multisample.count = key.msaa.samples();
+        descriptor.layout[0] = if key.msaa.samples() > 0 {
+            self.layout_msaa.clone()
+        } else {
+            self.layout.clone()
+        };
+        descriptor.fragment.targets[0].as_mut().unwrap().blend_mode = key.blend_state;
+        Ok(key)
+    }
+}
+
+impl GetBaseDescriptor for MySpecializer {
+    fn get_base_descriptor(&self) -> RenderPipelineDescriptor {
+        RenderPipelineDescriptor {
+            label: Some("my_pipeline".into()),
+            layout: vec![self.layout.clone()],
+            push_constant_ranges: vec![],
+            vertex: VertexState {
+                shader: self.vertex.clone(),
+                shader_defs: vec![],
+                entry_point: "vertex".into(),
+                buffers: vec![],
+            },
+            primitive: Default::default(),
+            depth_stencil: None,
+            multisample: MultiSampleState::default(),
+            fragment: Some(FragmentState {
+                shader: self.fragment.clone(),
+                shader_defs: vec![],
+                entry_point: "fragment".into(),
```

Comment:
i feel like the base descriptor should probably be passed in a fully defaulted value and this should only have to set a handful of things on it. Especially once we have wgpu automatic entrypoint selection, it would be really nice to unify base and specialize so that you just write one type of function. you'd only have to set the frag and vert shader, and the layout. down the line, potentially even the layout could be introspected from the shader and you wouldnt have to set that either

Reply:
This also came to mind when implementing. There isn't actually that much you can usefully set on a "base" descriptor besides perhaps a label, but that's also easy to set elsewhere as well.

Reply:
In fact, I kind of feel like this is the right call and would help simplify the API. Thoughts, @ecoskey?